kmip/pie/client.py

```diff
@@ -0,0 +1,389 @@
+
+        Args:
+            algorithm (CryptographicAlgorithm): An enumeration defining the
+                algorithm to use to generate the key pair.
+            length (int): The length in bits for the key pair.
+
+        Returns:
+            string: The uid of the newly created public key.
+            string: The uid of the newly created private key.
+
+        Raises:
+            ClientConnectionNotOpen: if the client connection is unusable
+            KmipOperationFailure: if the operation result is a failure
+            TypeError: if the input arguments are invalid
+        """
+        # Check inputs
+        if not isinstance(algorithm, enums.CryptographicAlgorithm):
+            raise TypeError(
+                "algorithm must be a CryptographicAlgorithm enumeration")
+        elif not isinstance(length, six.integer_types) or length <= 0:
+            raise TypeError("length must be a positive integer")
+
+        # Verify that operations can be given at this time
+        if not self._is_open:
+            raise exceptions.ClientConnectionNotOpen()
+
+        # Create the template containing the attributes
+        attributes = self._build_key_attributes(algorithm, length)
+        template = cobjects.CommonTemplateAttribute(attributes=attributes)
+
+        # Create the asymmetric key pair and handle the results
+        result = self.proxy.create_key_pair(common_template_attribute=template)
+
+        status = result.result_status.enum
+        if status == enums.ResultStatus.SUCCESS:
+            public_uid = result.public_key_uuid.value
+            private_uid = result.private_key_uuid.value
+            return public_uid, private_uid
+        else:
+            reason = result.result_reason.enum
+            message = result.result_message.value
+            raise exceptions.KmipOperationFailure(status, reason, message)
+
+    @_synchronize
+    def register(self, managed_object):
+        """
+        Register a managed object with a KMIP appliance.
+
+        Args:
+            managed_object (ManagedObject): A managed object to register. An
+                instantiatable subclass of ManagedObject from the Pie API.
+
+        Returns:
+            string: The uid of the newly registered managed object.
+
+        Raises:
+            ClientConnectionNotOpen: if the client connection is unusable
+            KmipOperationFailure: if the operation result is a failure
+            TypeError: if the input argument is invalid
+        """
+        # Check input
+        if not isinstance(managed_object, pobjects.ManagedObject):
+            raise TypeError("managed object must be a Pie ManagedObject")
+
+        # Verify that operations can be given at this time
+        if not self._is_open:
+            raise exceptions.ClientConnectionNotOpen()
+
+        # Extract and create attributes
+        attributes = list()
+        if hasattr(managed_object, 'cryptographic_usage_masks'):
```

Comment:
This is ok for now. In the future it would be cooler to not have to check hasattr for each attribute for the request. We can brainstorm this after this has been merged.